raiden/transfer/state.py

```diff
@@ -146,6 +158,195 @@ def __ne__(self, other):
+            self.nonce == other.nonce and
+            self.transferred_amount == other.transferred_amount and
+            self.locksroot == other.locksroot and
+            self.channel_address == other.channel_address
+        )
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class BalanceProofState2(State):
+    """ Proof of a channel balance that can be used on-chain to resolve
+    disputes.
+    """
+
+    __slots__ = (
+        'nonce',
+        'transferred_amount',
+        'locksroot',
+        'channel_address',
+        'message_hash',
+        'signature',
+        'sender',
+    )
+
+    def __init__(
+            self,
+            nonce: int,
+            transferred_amount: typing.token_amount,
+            locksroot: typing.keccak256,
+            channel_address: typing.address,
+            message_hash: typing.keccak256,
+            signature: typing.signature,
+            sender: typing.address):
+
+        if not isinstance(nonce, int):
+            raise ValueError('nonce must be int')
+
+        if not isinstance(transferred_amount, typing.token_amount):
+            raise ValueError('transferred_amount must be a token_amount instance')
+
+        if not isinstance(locksroot, typing.keccak256):
+            raise ValueError('locksroot must be a keccak256 instance')
+
+        if not isinstance(channel_address, typing.address):
+            raise ValueError('channel_address must be an address instance')
+
+        if not isinstance(message_hash, typing.keccak256):
+            raise ValueError('message_hash must be a keccak256 instance')
+
+        if not isinstance(signature, typing.signature):
+            raise ValueError('signature must be an signature instance')
+
+        if not isinstance(sender, typing.address):
+            raise ValueError('sender must be an address instance')
+
+        if nonce <= 0:
+            raise ValueError('nonce cannot be zero or negative')
+
+        if nonce >= 2 ** 64:
+            raise ValueError('nonce is too large')
+
+        if transferred_amount < 0:
+            raise ValueError('transferred_amount cannot be negative')
+
+        if transferred_amount >= 2 ** 256:
+            raise ValueError('transferred_amount is too large')
+
+        if len(locksroot) != 32:
+            raise ValueError('locksroot must have length 32')
+
+        if len(channel_address) != 20:
+            raise ValueError('channel is an invalid address')
+
+        if len(message_hash) != 32:
+            raise ValueError('message_hash is an invalid hash')
+
+        if len(signature) != 65:
+            raise ValueError('signature is an invalid signature')
+
+        self.nonce = nonce
+        self.transferred_amount = transferred_amount
+        self.locksroot = locksroot
+        self.channel_address = channel_address
+        self.message_hash = message_hash
+        self.signature = signature
+        self.sender = sender
+
+    def __str__(self):
```

Comment:
This (and in the other class) feels more like `__repr__` than `__str__`

Reply:
https://docs.python.org/3/reference/datamodel.html#object.__str__
> This method differs from object.__repr__() in that there is no expectation that __str__() return a valid Python expression: a more convenient or concise representation can be used.
`__repr__` should usually return a valid python expression, when executed instantiates and equivalente object

Reply:
It is not required to do so. Also coming from the doc you linked about `__repr__`:

>  If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment). If this is not possible, a string of the form <...some useful description...> should be returned. The return value must be a string object

I generally think of `__str__` as something to be consumed by users and `__repr__` by logs or developers.

And I am mentioning this since if you grep for `__repr__` in our codebase you wll find representations of a lot of objects looking much like the `__str__` you added here.

Reply:
> It is not required to do so. Also coming from the doc you linked about __repr__

True

> I generally think of __str__ as something to be consumed by users and __repr__ by logs or developers.

Our target audience is developers

> And I am mentioning this since if you grep for __repr__ in our codebase you wll find representations of a lot of objects looking much like the __str__ you added here.

This only means we are not using it properly.

Anyways, this is bikeshedding, will change it because it does not matter